### DISCUSSION TITLE
What's the easiest way to get a list of all property names of a Serverpod model?

### DISCUSSION THREAD

**Discussion by Frank1016** [serverpod/serverpod#1898] [Q&A]
Before updating to Serverpod 1.2, I can use one line of code to get a list of all property names of a model defined in server protocol folder like following:
`extension UserExtension on User {
  List<String> get properties => toJson().keys.toList();}`

Now I'm on Serverpod 1.2.3, and just found that this won't work any more. If I need to get a list of all property names of a model in Flutter app coding, do I have to manually create one now?

**Comment by Isakdl** [comment]
This should still work what is the issue you have?

Keep in mind that the method `toJson` will only serialize fields that are scoped to both the client and the server. If you restrict a field to the server only it will not be serialized with `toJson`. There is another method called `allToJson` which serializes all fields regardless of scope.

**Comment by Frank1016** [comment]
No, it doesn't work the same way now. All serialize fields are set default with scope to both client and server. Before Serverpod version 1.2, with the following code in the flutter folder: 
`print(List<String>.from(User().properties));`

I can get a list of all defined property names in User() class, but now with Serverpod 1.2.3, the same code will always get an empty list.

**Reply by Frank1016** [reply]
ChatGPT gives me the following suggestion:

> Instantiation of Abstract Class: Your example uses User().properties, implying you are trying to instantiate User directly. However, User is an abstract class and cannot be instantiated. This code should not compile unless there's a concrete implementation of User that is being instantiated. Ensure that the instantiation is of a concrete class that properly implements toJson(). 

Is this the reason why the same code won't work any more with Severpod 1.2?

**Reply by Isakdl** [reply]
Ahh I see, yes there was a PR we merge that made fields that are set to null not be serialized when converted to json, with the reasoning being to optimize the network traffic and serialization effort. 

This means if the field in the class is null the key will not be contained in the map, you could give each field a value and then serialize and you will get all keys in the toJson method.

May I ask you what your use-case is? Why do you need to extract the fields as strings?

**Reply by Frank1016** [reply]
What I'm doing is to create a form with basically all the fields from server-modeled Class and each property name is the key to identify each field. With the list of property names, I can use a for-loop to create all the textFormFields easily. 

Really don't want to create the list of property names manually for all server classes, as some of them have quite a lot of fileds defined, easily several hundreds.

**Reply by Isakdl** [reply]
Got it, makes sense, how do you deal with different datatypes though? Or do you simply only use this for strings?

Perhaps a better solution for this problem would be to get both the field name and the expected type? Kind of like a json schema.

**Reply by Frank1016** [reply]
You are right, regarding the datatypes, I still need to list all propery names manually like the following:
![image](https://github.com/serverpod/serverpod/assets/32797520/70445f31-807c-4d1c-8343-903bf89c4437)
It's not pretty, and what you just mentioned: "a json schema with both field name and data type" sounds to be a much better solution. Could you illustrate more on this in details? I'm really interested.

**Reply by Isakdl** [reply]
I'm just spitballing here, but maybe we could generate a method that would return a JSON schema (https://json-schema.org/learn/getting-started-step-by-step). This schema would fully represent the dart object with field names, nullability, and data types, etc.

This could then be used to create forms automatically in Flutter. For example, I found this package when I quickly searched pub.dev: https://pub.dev/packages/flutter_json_schema_form

A simpler version could be to return a map like:

```json
{
   "field": "String",
   "field2": "int?",
}
```

But that would still imply some parsing to figure out the types. 

In theory, you could parse the yaml files (models) as this information is in there already, but that would be a little cumbersome to implement I think as you would more or less have to replicate what the serverpod cli does.

I'll think about this, I think it is a pretty interesting use case, and it might make sense to have first-class support.

**Reply by Frank1016** [reply]
I see, thank you very much for the consideration! Really hope soon we can try this frist-class support from Serverpod!
